neutrons/SNAPRed
# Implement Create Artificial Norm Algo within Reduction Workflow

## Description of work
The purpose for this PR is the implement the previously created artificial normalization within the `Reduction` workflow. The changes ensure the workflow can handle missing normalization data by applying artificial normalization and introduce enhancements to the user interface, logging, and data handling processes.


## Explanation of work
Key updates include:

1. Artificial Normalization:
* Added the `ArtificialNormalizationRecipe` to `GenericRecipe` that uses the `CreateArtificialNormalizationAlgo` to handle cases when normalization data is missing.
* Automatically extracts parameters and queues the algorithm to create an artificial normalization workspace when needed.

2. Workflow Updates:
* The `Reduction` workflow can now dynamically adjust for missing calibration or normalization data, applying artificial normalization when necessary.
* Updated the UI to prompt users for artificial normalization inputs and seamlessly continue the workflow.

3. Error Handling and Logging:
* Improved logging and error messages to guide users through situations where data is missing.
* Enhanced exception handling to gracefully manage missing states or permissions.

4. Backend Improvements:
* Refactored the `ReductionService` to handle normalization and calibration more flexibly.
* Expanded unit tests to cover new cases with artificial normalization.



## To test
Insure all pytests pass. Please launch SNAPRed UI with a fresh (no states initialized) local copy of `Calibration_next` specified within `application.yml`. `First` run the Diffraction Calibration with the following inputs:

* Run Number: 58882
* Convergence Threshold: 0.1
* Bins Across Peak Width: 10
* Sample: Silicon_NIST_640D_001.json
* Grouping File: Column

Create a new state when prompted. Then continue on through workflow through successful completion.
Now moving onto the Reduction tab, notice there is a new tab called "Artificial Normalization". Press the `Continue` button and notice the UI handling for no inputs:

![image](https://github.com/user-attachments/assets/6ca7f6b9-1f45-4419-9107-3c6aeaaa200e)

 Finally, `without` running a normalization, run the `Reduction` workflow with the following inputs:

* Run Number: 58882

Continue the workflow to notice the following message pop-up:

![image](https://github.com/user-attachments/assets/7671b1bb-56a3-434b-81ad-95a6e8f979cd)

Continue Past the possible warning for permissions check but clicking `Yes`. Find that the workflow completes successfully. 

`Finally`, re-run this but with a normalization this time and see that the `Artificial Normalization` flow is not used in this case.
### Dev testing
Insure all pytests pass


### CIS testing
See above.




## Link to EWM item


[EWM # 7123](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20(Change%20Management)#action=com.ibm.team.workitem.viewWorkItem&id=7123)

&

[EWM # 7121](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20(Change%20Management)#action=com.ibm.team.workitem.viewWorkItem&id=7121)

### Verification




### Acceptance Criteria



This list is for ease of reference, and does not replace reading the EWM story as part of the review.  Verify this list matches the EWM story before reviewing.

- **Note** - No acceptance criteria is provided for this story. This recipe shall be fully testable within the acceptance criteria of the UI portion of this EPIC.
